Explicitly make dynamically dispatched API !Send + !Sync

mitchmindtree opened the pull request:
This is in order to ensure consistent restrictions across platforms in a
manner that ensures thread safety across each of the supported
platforms.

Please see added comments in the diff for details on which platforms
impose these restrictions.

@ishitatsuyuki is this kind of what you had in mind? I initially tried using negative trait bounds but it seems they're not yet stable.

mitchmindtree committed:
Explicitly make dynamically dispatched API !Send + !Sync

This is in order to ensure consistent restrictions across platforms in a
manner that ensures thread safety across each of the supported
platforms.

Please see added comments in the diff for details on which platforms
impose these restrictions.

ishitatsuyuki commented on src/platform/mod.rs:
```diff
@@ -48,19 +48,31 @@ macro_rules! impl_platform_host {
+        // TODO: Try to contact ASIO to get more information. Review the existing implementation of
+        // the `asio` backend's `Host` and `Driver` types and see if the existing `Arc`s and
+        // `Mutex`s don't already make the API `Sync`.
+        pub struct Host(HostInner, crate::platform::NotSyncAcrossAllPlatforms);
```
Hmm, so I think this should be one of the following instead:
- `!Send` and `!Sync` because we use Arc and the handles end up being shared between `Host` and `Device`s
- Just assume it's thread safe because we already wrap almost everything inside a Mutex.

mitchmindtree replied:
After a closer review it seems that the higher-level API exposed via the asio crate does indeed force synchronised access to the API and state transitions via the `Mutex` - I noticed some parts that weren't covered properly and patched them up.

ishitatsuyuki reviewed (approved):
I think I was wrong on the ASIO part, see inline comments. Otherwise this looks good! Thanks.

mitchmindtree committed:
Fix state transition synchronisation in ASIO

This makes some tweaks to the ASIO backend in order to fix some cases
where races may have occured. This should allow us to remove the `Sync`
bound on the `Device` and `Host` types.

mitchmindtree committed:
Relax `Sync` restriction on `Device` and `Host`

Originally this restriction was placed due to uncertainty around the
thread safety of the ASIO API. While the ASIO API itself makes no
thread-safety guarantees that we are aware of, the `asio-sys` high-level
bindings enforce synchronised access to the API and state transitions
via a mutex.
